### PR TITLE
Improve device info with new flops and bandwidth formula based on hardware libraries

### PR DESCRIPTION
**Pull request by exclamaforte**
Previously, DeviceInfo provided theoretical hardware information based on a hardcoded list manually created from various datasheets. 

This update:
- Attempting to gather the information from a hardware library like `pynvml`, improving accuracy and expanding support to devices that don't have entries in the datasheet list.
- Adjusts flops and bw calculation based on these hardware values. For example, if the the memory or SMs are underclocked, it adjusts the theoretical max flops/bw accordingly.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

### CODE REVIEW

File: torch/_inductor/analysis/device_info.py [hunk 1]
```diff
@@ -1,39 +1,374 @@
...
+    def lookup_memory_clock_hz(device_name: str) -> Optional[float]:
+        """Get the memory clock speed in Hz for the current device."""
+        result = DeviceInfo._generic_lookup(device_name, "memory_clock_hz")
+        return result if isinstance(result, (int, float)) or result is None else None
+
+    @staticmethod
+    def lookup_dram_gb(device_name: str) -> Optional[float]:
+        """Get the DRAM memory size in GB for the current device."""
+        result = DeviceInfo._generic_lookup(device_name, "dram_gb")
+        return result if isinstance(result, (int, float)) or result is None else None
```
Comment (shunting314) [inline torch/_inductor/analysis/device_info.py]: whey need this rather than 'return result' directly?  Same for multiple other similar methods.

File: torch/_inductor/analysis/device_info.py [hunk 2]
```diff
@@ -1,39 +1,374 @@
...
+        if amd_smi is None:
+            return None
+
+        try:
+            amd_smi.amdsmi_init()
+            device_handle = amd_smi.amdsmi_get_processor_handles()[0]
+            mem_clock_info = amd_smi.amdsmi_get_clock_info(
+                device_handle, amd_smi.AmdSmiClkType.MEM
+            )
+            amd_smi.amdsmi_shut_down()
```
Comment (shunting314) [inline torch/_inductor/analysis/device_info.py]: Create a common function to lookup hw spec via amd smi.

Same for pynvml case.

File: torch/_inductor/analysis/device_info.py [hunk 3]
```diff
@@ -1,39 +1,374 @@
...
+from typing import Any, Optional, Union
 
 import torch
 
 
 log = logging.getLogger(__name__)
 
 
+# Lazy import cache for pynvml
+_pynvml_cache: Optional[Any] = None
```
Comment (v0i0) [inline torch/_inductor/analysis/device_info.py]: torch._cuda._HAS_PYNVML etc seem to already do this for you, why not use that?

File: torch/_inductor/analysis/device_info.py [hunk 4]
```diff
@@ -1,39 +1,374 @@
...
+        """Get the number of streaming multiprocessors from the hardware."""
+        try:
+            # rely on device_properties api
+            device_props = torch.cuda.get_device_properties(0)
+            return device_props.multi_processor_count
+        except Exception:
+            return None
+
+    @staticmethod
+    def _hardware_lookup_clock_hz() -> Optional[float]:
```
Comment (v0i0) [inline torch/_inductor/analysis/device_info.py:124]: why are we using nvml vs a device query for that? is the idea that this might capture current frequency (as opposed to the configured application clock)?

File: torch/_inductor/analysis/device_info.py [hunk 5]
```diff
@@ -1,39 +1,374 @@
...
 # Indexing is based on `torch.cuda.get_device_name()`
 # TODO investigate profiler support for tf32 and allow device to report correct number when it's turned on.
-_device_mapping: dict[str, DeviceInfo] = {
+_device_mapping: dict[str, DeviceSpec] = {
     # Source:
     # @lint-ignore https://www.nvidia.com/en-us/data-center/h100/
-    "NVIDIA H100": DeviceInfo(
+    # These are from H100 SXM.
+    #
+    "NVIDIA H100": DeviceSpec(
```
Comment (v0i0) [inline torch/_inductor/analysis/device_info.py:362]: i wonder what it would take (in terms of making scaling these up top e.g. by SM in addition to frequency) to support the entirety of the Hopper device "zoo"

**Review by v0i0 — APPROVED** [review]
looks great, left some nits / questions

File: torch/_inductor/analysis/device_info.py (hunk 4 above)
Comment (exclamaforte) [reply torch/_inductor/analysis/device_info.py:124]: I changed it to nvmlDeviceGetMaxClockInfo, I think that's the best bet

File: torch/_inductor/analysis/device_info.py (hunk 5 above)
Comment (exclamaforte) [reply torch/_inductor/analysis/device_info.py:362]: it would be pretty easy refactor because you can query sm count, I'll do a followup PR.

**Review by shunting314 — APPROVED** [review]

**Review by eellison — COMMENTED** [review]
deferring to @v0i0 and @shunting314 but great to see this change!

File: torch/_inductor/analysis/device_info.py [hunk 6]
```diff
@@ -171,6 +526,8 @@ def datasheet_tops(dtype: torch.dtype, is_tf32: bool = False) -> Optional[float]
     Get the theoretical TFLOPS of the device for a given dtype. This can throw an exception if the device
     is not in the datasheet list above.
     """
+    if not torch.cuda.is_available():
```
Comment (xmfan) [inline torch/_inductor/analysis/device_info.py]: is this check needed? _device_mapping doesn't contain non nvidia/amd, and if it did shouldn't we still return their entry
